gb-joypad/src/register.rs

```diff
@@ -16,27 +19,135 @@ struct RegisterBits {
     void: B2,
 }
 
+#[derive(Debug, Clone, Copy)]
 enum RegisterMode {
     Direction,
     Action,
+    Unset,
 }
 
+#[derive(Default, Debug, Clone, Copy)]
 pub struct JoypadRegister {
     bits: RegisterBits,
     mode: RegisterMode,
 }
 
+impl RegisterBits {
+    fn update(
+        &mut self,
+        values: [InputType; 4],
+        state: &mut HashMap<InputType, bool>,
+        addr_bus: &mut dyn Bus<u8>,
+    ) {
+        let old = self.p10();
+        self.set_p10((!state[&values[0]]).into());
+        if old != 0 && self.p10() == 0 {
+            trigger_interrupt(addr_bus);
+        };
+
+        let old = self.p11();
+        self.set_p11((!state[&values[1]]).into());
+        if old != 0 && self.p11() == 0 {
+            trigger_interrupt(addr_bus);
+        };
+
+        let old = self.p12();
+        self.set_p12((!state[&values[2]]).into());
+        if old != 0 && self.p12() == 0 {
+            trigger_interrupt(addr_bus);
+        };
+
+        let old = self.p13();
+        self.set_p13((!state[&values[3]]).into());
+        if old != 0 && self.p13() == 0 {
+            trigger_interrupt(addr_bus);
+        };
+    }
+}
+
+fn trigger_interrupt(addr_bus: &mut dyn Bus<u8>) {
+    let interrupts_val = addr_bus
+        .read(0xFF00, None)
+        .expect("Failed to read interrupt value for joypad interrupt");
+    if let Err(err) = addr_bus.write(0xFF00, interrupts_val | 0b10000, None) {
```

Comment:
The use of a constant here for the address would probably be more readable. Plus I think that this is the wrong address to access the interrupts flag.

Reply:
You're right the address was wrong.